### ISSUE TITLE
Obspy merge function in TimeseriesUtility

### ISSUE THREAD

**Issue by arigdon-usgs** [usgs/geomag-algorithms#216]
In TimeseriesUtility.merge_streams() the obspy merge function (shown in link below) fails for a specific instance of a certain type of gap. The issue is located within obspy and numpy as y.astype(dtype,copy=False) is called. The issue is that the built-in astype function for a numpy.ndarray contains a keyword for copy but the built-in astype function for a numpy.ma.MaskedArray does NOT contain a keyword for copy. While the astype is called within the numpy library it is only called for a specific instance from obspy. Obspy calls the linspace in the obspy trace.__add__ function when delta is < 0, delta endtime < 0, method = 1, and (-1 <= interpolation_samples < delta).

My understanding is that obspy is trying to create a linspace to fill the empty gaps of the masked array. While trying to fill the gap it is grabbing the last empty data point and filling from there. That last empty point is a masked point and causes linspace to fail. If it grabbed the first real data point it would pass and it run. I'm not entirely sure why this issue has only just begun to show up.

https://github.com/usgs/geomag-algorithms/blob/93987f20f9534042401dd991a06710967903fe18/geomagio/TimeseriesUtility.py#L222

**Comment by jmfee-usgs** [comment]
Do you have a workaround for this?  If need be, we could add this to this project for now, and submit a patch to the upstream obspy project for eventual inclusion; or, is it easier to detect these streams that cause problems and not include them in the stream being merged?

**Comment by arigdon-usgs** [comment]
I have not been able to find a stable workaround for this or a good way of determining which stream will fail. Although this does seem to be mainly an issue on the seconds data and has not affected minutes data.

**Comment by erigler-usgs** [comment]
To be consistent with the current TimeseriesUtility logic, which puts forth considerable effort to break  series into contiguous, unmasked segments before merging them, a fix should ultimately be applied to ObsPy and/or NumPy's masked array code.

HOWEVER...I think I have a simple work-around:

If we use the optional `fill_value` keyword for Stream.merge(), we can address the underlying problem, which is that ObsPy recursively builds up traces with segment traces with identical IDs. If it finds a gap, it either changes the trace to a masked array, or fills the gap with `fill_value`. If the former (default), the interpolation logic in Trace.\_\_add\_\_() breaks (or more accurately, numpy.linspace() breaks, but that seems less likely to get fixed than ObsPy).

In short, the following change to TimeseriesUtility.merge_streams() seems to fix the problem:

```python
    # merge data
    split.merge(
            # 1 = do not interpolate
            interpolation_samples=0,
            # 1 = when there is overlap, use data from trace with last endtime
            method=1,
            # np.nan = work-around for (problematic) intermediate masked arrays in Stream.merge()
            filled_value=numpy.nan)
```

I tested this manually, and it seems to work as expected. I admit I haven't thought this change through super carefully, but I don't believe there would be any side-effects beyond some possibly unnecessary NaN calculations (which simply result in another NaN, which we convert anyway in TimeseriesUtility).

Thoughts?

**Comment by aclaycomb-usgs** [comment]
I like this plan.

**Comment by erigler-usgs** [comment]
I wrapped this tiny work-around into PR #217 . Closing this issue.
